File: backend/src/main/java/com/rentacar/backend/services/UsuarioService.java
package com.rentacar.backend.services;

import com.rentacar.backend.dto.UsuarioDTO;
import com.rentacar.backend.entities.*;
import com.rentacar.backend.repositories.ReservaRepository;
import com.rentacar.backend.repositories.SucursalRepository;
import com.rentacar.backend.repositories.UsuarioRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.security.core.userdetails.UserDetails;
import org.springframework.security.core.userdetails.UserDetailsService;
import org.springframework.security.core.userdetails.UsernameNotFoundException;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;
import java.util.NoSuchElementException;
import java.util.Optional;
import java.util.stream.Collectors;

@Service
@Transactional
public class UsuarioService implements UserDetailsService {

    private final UsuarioRepository usuarioRepository;
    private final ReservaRepository reservaRepository;
    private final SucursalRepository sucursalRepository;
    private final PasswordEncoder passwordEncoder;

    @Autowired
    public UsuarioService(UsuarioRepository usuarioRepository, ReservaRepository reservaRepository, SucursalRepository sucursalRepository, PasswordEncoder passwordEncoder) {
        this.usuarioRepository = usuarioRepository;
        this.reservaRepository = reservaRepository;
        this.sucursalRepository = sucursalRepository;
        this.passwordEncoder = passwordEncoder;
    }

    public UsuarioEntity obtenerUsuarioPorId(Long id) {
        Optional<UsuarioEntity> usuario = usuarioRepository.findById(id);
        if (usuario.isEmpty()) throw new NoSuchElementException("Usuario no encontrado");
        return usuario.get();
    }


    // Actualizar usuario que ya existe, por id
    public UsuarioEntity actualizarUsuario(Long id, UsuarioDTO usuarioDTO) {
        Optional<UsuarioEntity> usuarioExistente = usuarioRepository.findById(id);

        if (usuarioExistente.isEmpty()) {
            throw new RuntimeException("Usuario no encontrado con ID: " + id);
        }

        UsuarioEntity usuario = usuarioExistente.get();

        if (usuarioDTO.getNombre() != null) {
            usuario.setNombre(usuarioDTO.getNombre());
        }
        if (usuarioDTO.getApellido() != null) {
            usuario.setApellido(usuarioDTO.getApellido());
        }
        if (usuarioDTO.getPassword() != null) {
            usuario.setPassword(usuarioDTO.getPassword());
        }
        if (usuarioDTO.getRol() != null) {
            usuario.setRol(usuarioDTO.getRol());
        }

        // Handle sucursal assignment
        if (usuarioDTO.getSucursalId() != null) {
            SucursalEntity sucursal = sucursalRepository.findById(usuarioDTO.getSucursalId())
                    .orElseThrow(() -> new RuntimeException("Sucursal no encontrada con ID: " + usuarioDTO.getSucursalId()));

            //usuario.setSucursal(sucursal);
            sucursal.getEmpleados().add(usuario);
            sucursalRepository.save(sucursal);
        }

        usuario.setEstaEnListaNegra(usuarioDTO.isEstaEnListaNegra());

        return usuarioRepository.save(usuario);
    }

    // Eliminar usuario por id
    public void eliminarUsuario(Long id) {
        Optional<UsuarioEntity> usuarioOptional = usuarioRepository.findById(id);

        if (usuarioOptional.isEmpty()) {
            throw new RuntimeException("Usuario no encontrado con ID: " + id);
        }

        UsuarioEntity usuario = usuarioOptional.get();


        // Check if user has active reservations
        List<ReservaEntity> reservasActivas = reservaRepository.findByUsuario(usuario)
                .stream()
                .filter(r -> r.getEstado() == ReservaEntity.EstadoReserva.EN_PROGRESO ||
                        r.getEstado() == ReservaEntity.EstadoReserva.CONFIRMADA)
                .toList();

        if (!reservasActivas.isEmpty()) {
            throw new RuntimeException("No se puede eliminar un usuario con reservas activas");
        }

        try {
            usuario.setSoftDelete(true);
            usuarioRepository.save(usuario);

        } catch (Exception e) {
            throw new RuntimeException("Error al eliminar el usuario: " + e.getMessage());
        }
    }
    //Asignar sucursal a un usuario al momento de crear un usuario.
    public UsuarioEntity asignarSucursal(Long usuarioId, Long sucursalId) {
        UsuarioEntity usuario = usuarioRepository.findById(usuarioId)
                .orElseThrow(() -> new RuntimeException("Usuario no encontrado con ID: " + usuarioId));

        SucursalEntity sucursal = sucursalRepository.findById(sucursalId)
                .orElseThrow(() -> new RuntimeException("Sucursal no encontrada con ID: " + sucursalId));

        //usuario.setSucursal(sucursal);
        return usuarioRepository.save(usuario);
    }
    
    public List<UsuarioEntity> obtenerTrabajadores() {
        return usuarioRepository.findByRol(UsuarioEntity.RolUsuario.TRABAJADOR);
    }

    public List<UsuarioEntity> obtenerAdministradores() {
        return usuarioRepository.findByRol(UsuarioEntity.RolUsuario.ADMINISTRADOR);
    }

    public List<UsuarioEntity> obtenerArrendatarios() {
        return usuarioRepository.findByRol(UsuarioEntity.RolUsuario.ARRENDATARIO);
    }

    public List<UsuarioEntity> obtenerUsuariosActivos() {
        return usuarioRepository.findBysoftDelete(false);
    }

    @Override
    public UserDetails loadUserByUsername(String rut)
        throws UsernameNotFoundException {
        return usuarioRepository.findByRut(rut)
            .orElseThrow(() ->
                             new UsernameNotFoundException("Usuario no encontrado"));
    }

    public UsuarioEntity crearUsuario(String rut, String nombre,
                                      String apellido, String password, UsuarioEntity.RolUsuario rol) {
        if (usuarioRepository.findByRut(rut).isPresent()) {
            throw new RuntimeException("Usuario ya existe");
        }

        UsuarioEntity usuario = new UsuarioEntity();
        usuario.setRut(rut);
        usuario.setNombre(nombre);
        usuario.setApellido(apellido);
        usuario.setPassword(passwordEncoder.encode(password));
        usuario.setRol(rol);
        return usuarioRepository.save(usuario);
    }

    public Optional<UsuarioEntity> validarCredenciales(
        String rut, String password, String rol) {
        Optional<UsuarioEntity> usuario = usuarioRepository.findByRut(rut);

        if (usuario.isPresent() &&
            passwordEncoder.matches(password, usuario.get().getPassword()) &&
            usuario.get().getRol().name().equals(rol) &&
            !usuario.get().isSoftDelete()) {
            return usuario;
        }

        return Optional.empty();
    }
}

File: frontend/src/services/UsuarioService.js
import axiosInstance from './axiosConfig';

export const useUsuarioService = () => {
  const crearUsuario = async (usuario) => {
    try {
      const {data} = await axiosInstance.post('/api/usuario/crear', {
        rut: usuario.rut,
        nombre: usuario.nombre,
        apellido: usuario.apellido,
        password: usuario.password,
        rol: usuario.rol,
        sucursalId: usuario.sucursalId
      });
      return data;
    } catch (error) {
      if (error.response) {
        switch (error.response.status) {
          case 400:
            throw new Error(error.response.data || 'Datos inválidos');
          default:
            throw new Error('Error en el servidor');
        }
      }
      throw error;
    }
  };

  const obtenerTrabajadores = async () => {
    try {
      const {data} = await axiosInstance.get('/api/usuario/trabajadores');
      return data;
    } catch (error) {
      throw new Error(error.response?.data || 'Error al obtener trabajadores');
    }
  };

  const obtenerAdministradores = async () => {
    try {
      const {data} = await axiosInstance.get('/api/usuario/administradores');
      return data;
    } catch (error) {
      throw new Error(error.response?.data || 'Error al obtener administradores');
    }
  };

  const obtenerArrendatarios = async () => {
    try {
      const {data} = await axiosInstance.get('/api/usuario/arrendatarios');
      return data;
    } catch (error) {
      throw new Error(error.response?.data || 'Error al obtener arrendatarios');
    }
  };

  const eliminarUsuario = async (id) => {
    try {
      const response = await axiosInstance.delete(`/api/usuario/eliminar`, {
        params: { id }
      });

      if (response.status === 200) {
        return response.data;
      }

      throw new Error(response.data?.error || 'Error al eliminar usuario');
    } catch (error) {
      if (error.response?.data?.error) {
        throw new Error(error.response.data.error);
      }
      throw new Error('Error al eliminar usuario');
    }
  };

  const actualizarUsuario = async (id, userData) => {
    try {
      const usuarioDTO = {
        nombre: userData.nombre,
        apellido: userData.apellido,
        rol: userData.rol,
        estaEnListaNegra: userData.estaEnListaNegra
      };

      const { data } = await axiosInstance.put(`/api/usuario/actualizar?id=${id}`, usuarioDTO);
      return data;
    } catch (error) {
      throw new Error(error.response?.data || 'Error al actualizar usuario');
    }
  };

  const login = async (rut, password, rol) => {
    try {
      const {data} = await axiosInstance.post('/api/usuario/login', {
        rut,
        password,
        rol
      });
      return data;
    } catch (error) {
      if (error.response) {
        switch (error.response.status) {
          case 403:
            throw new Error('Usuario en lista negra');
          case 401:
            throw new Error('Credenciales inválidas o rol no autorizado');
          default:
            throw new Error('Error en el servidor');
        }
      }
      throw error;
    }
  };

  return {
    login,
    crearUsuario,
    obtenerTrabajadores,
    obtenerAdministradores,
    obtenerArrendatarios,
    eliminarUsuario,
    actualizarUsuario
  };
};
